File: Project_Euler/21-Amicable_Numbers/Amicable_Numbers.py
def sum_divisors(n):
    # return the sum of proper divisors of n
    
    res = 1
    
    for i in range(2,1+int(n**0.5)):
        if n%i==0:
            res+=i
            if n!=i*i: # if proper square, just add once
                res+=int(n/i)
        
    return res

sum_of_divisors = [0]*10000 # store sum of divisors

for i in range(2,10000):
    sum_of_divisors[i] = sum_divisors(i)

ans = 0
for i in range(2,10000):
    if sum_of_divisors[i]<10000 and sum_of_divisors[i]!=i and sum_of_divisors[sum_of_divisors[i]]==i:
        ans += i

print(ans)

File: Project_Euler/21-Amicable_Numbers/Amicable_Numbers.cpp
#include <iostream>
#include <math.h>
using namespace std;

long long int sum_div(int n) {
    long long int res = 1;
    int lim = (1 + (int)(sqrt(n)));
    for (int i = 2; i < lim; i++) {
        if (!(n % i)) {
            res =res + i;
            if (n != (i * i)) {
                res =res + (int)(n / i);
            }
        }
    }
    
    return res;
}
int main() {	
	long long int sum_of_divisors[10000];
    for (int i = 2; i < 10000; i++) {
        sum_of_divisors[i] = sum_div(i);
    }
    
    long long int ans;
    for (int i = 2; i < 10000; i++) {
        if (sum_of_divisors[i] < 10000 && sum_of_divisors[i] != i && sum_of_divisors[sum_of_divisors[i]] == i) {
            ans = ans + i;
        }
    }
    cout<<ans<<endl;
	return 0;
}
